p2p/src/protocol/txnmgr.rs

```diff
@@ -0,0 +1,92 @@
+//!
+//! Manages transaction confirmation.
+//!
+#![warn(missing_docs)]
+use std::collections::HashMap;
+
+use bitcoin::{network::message::NetworkMessage, Transaction, Txid};
+
+use crate::PeerId;
+
+use super::channel::Channel;
+
+/// The status of a transaction.
+#[derive(Clone, Debug)]
+pub enum Event {
+    /// The transaction was sent to one or more peers on the network, and is unconfirmed.
+    Pending {
+        /// Number of announcements of this transaction from other peers.
+        announcements: u8,
+        /// Transaction hash.
+        txid: Txid,
+    },
+}
+
+impl std::fmt::Display for Event {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Event::Pending {
+                announcements,
+                txid,
+            } => {
+                write!(
+                    fmt,
+                    "Pending transaction {} broadcasted to {} peers",
+                    txid, announcements
+                )
+            }
+        }
+    }
+}
+
+/// The ability to emit SPV related events.
+pub trait Events {
+    /// Emit an SPV-related event.
+    fn event(&self, event: Event);
+}
+
+/// Manages transactions.
+#[derive(Debug)]
+pub struct TxnManager<U> {
+    /// Transaction IDs to be tracked.
+    transactions: HashMap<Txid, (Transaction, u8)>,
+    upstream: U,
+}
```

Comment:
As mentioned on Discord, I think it makes more sense for the Client to own this map of transactions, because it's a user-created object and has no significance to the light client protocol. The hook on get_data is there so that the client can respond with the transaction payload when it is requested.

Reply:
For that reason I think this whole txnmgr should be a module in nakamoto-client instead. The key though is really that if the user wants to iterate over his transactions, these should be stored in the user thread, not in the protocol thread. A design that enables that will be more in line with the separation between client, wallet and protocol. In the Rust sense, the owner of the `Transaction` map should be the user of the library, or perhaps the Client, though that would require the Transactions to be in an `Arc<Mutex<_>>`.

Reply:
With this, the question of how the user gets to know that a tx is requested via get_data remains:

* It could be that Client wraps the get_data hook and keeps track of requested txids before proxying the call to the user-supplied hook
* It could be that the protocol emits an event on get_data and the client listens and keeps track of counts

Reply:
It may help to write some example code from the user perspective, and see how that could be implemented. Giving it a try:

```rust
let handle: client::Handle;
let tx1: Transaction;
let tx2: Transaction;

handle.submit_transaction(tx1);
handle.submit_transaction(tx2);

// I want to know when this tx has been accepted by N peers.

let txmgr = client::txmgr::TxManager::new(handle.clone());
let _ = txmgr.wait(tx1.id); // Wait for a change of status of tx1
// We can also do `txmgr.submit(tx1)` which does the above

// Iterate over transaction status events
for event in txmgr.events() {
  match event {
     client::txmgr::Event::TransactionAccepted(txid, peer_id) => {
         // This tx was accepted by a peer.
         // When enough peers accept a tx, do something ...
     }
     client::txmgr::Event::TransactionConfirmed(txid, confirmations) => { ... } // Included in a block
  }
}
```

Reply:
The above makes sense to me because transaction management involves elements relating to the wallet (cfilter matching) as well as the p2p network (getdata and inv). So it seems to not really belong in either, but as a separate thing.

Reply:
I'm not too sure how we are to track confirmed transactions since we can't get a `reject confirmed` status from peers.

Reply:
> I'm not too sure how we are to track confirmed transactions since we can't get a `reject confirmed` status from peers.

This has to be done by tracking cfilters in the txmgr, there's no other way.